[CIAB] BookingListViewModel refactoring and some fixes

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

### Description
This PR has three main changes:

- Implement the approach [discussed](https://github.com/woocommerce/woocommerce-android/pull/14735#discussion_r2420675966) previously to make the ViewModel more reactive, and avoid reading the fetch parameters from `StateFlow#value`.
- Fix an issue with pull-to-refresh and search (check the comments for more details)
- Fix an issue with the WCSearchView composable (check the comments for more details)

### Steps to reproduce
1. Use a CIAB site that has multiple bookings, multiple pages.
2. Open bookings list.

### Testing information
- Test changing tabs, searching.
- Test changing the sorting order, and confirm it shows the refresh status.
- Check for any regressions.

### The tests that have been performed
The above.

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Refactor BookingListViewModel logic to make it more reactive

hichamboushaba committed:
Make sure search results are not reset when refreshing

hichamboushaba committed:
Add unit tests

hichamboushaba committed:
Fix issue with not updating the search view when query is changed externally

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/list/BookingListHandler.kt:
```diff
@@ -61,14 +62,17 @@ class BookingListHandler @Inject constructor(
             fetchBookings()
         } else {
-            searchResults.value = emptyList()
+            if (searchQuery != previousQuery) {
+                searchResults.value = emptyList()
+            }
```
This is the main part of the second change mentioned in the description, it fixes an issue where pull-to-refresh would result in displaying the Empty state during the fetch. I added two unit tests to confirm the behavior too.

Check the below recordings to see the diffference:
| Before | After |
| --- | --- |
| <video src=https://github.com/user-attachments/assets/1fd5ef87-bace-4093-9553-d494f07077b1 /> | <video src=https://github.com/user-attachments/assets/159c6605-f070-42d2-a644-a4bae219b9fd /> |

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/TextFields.kt:
```diff
@@ -209,10 +209,13 @@ fun WCSearchField(
     var textFieldValueState by rememberSaveable(stateSaver = TextFieldValue.Saver) {
         mutableStateOf(TextFieldValue(text = value))
     }
+    // Keep TextFieldValue in sync with external value
+    val textFieldValue = textFieldValueState.copy(text = value)
```
Without this, when the value is changed externally (for example by tapping on the clear button), the update is not reflected on the text field.
This line is copied from the `BasicTextField` String implementation.

| Before | After |
| --- | --- |
| <video src=https://github.com/user-attachments/assets/ddc46601-95dc-4238-8e04-3a1477b28087 /> | <video src=https://github.com/user-attachments/assets/b30bda06-cd35-4b59-819f-74b9e03aa317 /> |

AdamGrzybkowski commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/list/BookingListHandler.kt:
```diff
@@ -48,7 +49,7 @@ class BookingListHandler @Inject constructor(
             )
         } else {
-            searchResults
+            searchResults.take(page * PAGE_SIZE)
```
Do we need that? 

`searchResults` returns a List of bookings, not single bookings so I'm not sure why we take `page * PAGE_SIZE` 🤔

hichamboushaba replied:
Given we don't clear the results now when after refresh if the query is not modified, in case more than a single page was previously fetched, this will mean that the "load more" login won't work as expected without this, as we won't reach the view that triggers it.
This is similar to what we do with the DB observation too.

Does this clarifies it?

AdamGrzybkowski replied:
To be honest, no 😅 

>This is similar to what we do with the DB observation too.

DB uses that value to limit the nr of entries returned. Here we are `taking` a specific number of values (list#1, list#2...) so those seem like different things 🤔 

I'm testing the code and if I remove the `take` operator it it behaves in the same way. I can reach the view that triggers the next page load.

Something like this would make more sense `searchResults.map { it.take(page * PAGE_SIZE) }` but I'm not sure we need that at all. When the pull to refresh is triggered the `page` is reset to 1, so we for the short moment when the refresh is happening we would show less items than before triggering it 🤔

hichamboushaba replied:
> `searchResults.map { it.take(page * PAGE_SIZE) }`

That's what I thought I was doing 🤦 😅 

> I'm testing the code and if I remove the take operator it it behaves in the same way. I can reach the view that triggers the next page load.

The question here, is what's the end of the list? If we keep all the items, then the end of the list is further more than our current page, this breaks the pagination logic, because when the first page is fetched, we'll then get to this [line](https://github.com/woocommerce/woocommerce-android/blob/6fe7f698d4520188876874083770452d614226a2/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/list/BookingListHandler.kt#L111) while will remove the items you are seeing at the end of list, and then the list will jump above, and won't be a good experience.
Again, this logic is similar to what we do with DB, because we clear the DB after the fetch, we ask the DB to return only the number of items that fits in our pages count.

Let me know if this still doesn't clarify it 🙂, if you agree with the above, then I'll fix the logic to apply `take` on the list of items and not the `Flow`.

AdamGrzybkowski replied:
>The question here, is what's the end of the list? If we keep all the items, then the end of the list is further more than our current page, this breaks the pagination logic, because when the first page is fetched, we'll then get to this [line](https://github.com/woocommerce/woocommerce-android/blob/6fe7f698d4520188876874083770452d614226a2/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/list/BookingListHandler.kt#L111) while will remove the items you are seeing at the end of list, and then the list will jump above, and won't be a good experience.

When you pull to refresh you are already at the top, so there's no visible jump in the list as there no visible change. 

Anyway

>Let me know if this still doesn't clarify it 🙂, if you agree with the above, then I'll fix the logic to apply take on the list of items and not the Flow.


That sounds good! My main issue was the `.take` on the Flow, which was breaking my brain, because I though I'm missing something obvious.

hichamboushaba replied:
> When you pull to refresh you are already at the top, so there's no visible jump in the list as there no visible change.

But the user can scroll down when the fetch is ongoing, and if the fetch takes long, then this would be noticeable.

Thanks @AdamGrzybkowski for catching the issue, I made the change in the last commit edcbf72

AdamGrzybkowski commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/list/BookingListViewModel.kt:
```diff
@@ -115,37 +116,66 @@ class BookingListViewModel @Inject constructor(
+                        initialLoadingState = BookingListLoadingState.Refreshing,
+                        fetchParams = fetchParams
+                    )
+                }
```
I think we are good here, but I tend to agree, I never like a collect inside another collect.

hichamboushaba replied:
Yes we are good, the `collectLatest` means that we will cancel the nested `collect` on each new emission.

The above patterns allows us to retrigger the fetching with the same parameters without reverting to reading the values from the `Flows` like how we did before, so I think it fits our need well.

hichamboushaba replied:
I ended up getting rid of the nested collection, as it caused a minor issue before: pull-to-refresh action needed to wait for any pending fetch to finish before triggerring, with the last change in the commit 090a1489b23eea2c049e8c5304b5da664df5b1f6, we will cancel pending fetches on pull-to-refresh, which matches the previous behavior.

AdamGrzybkowski reviewed (changes requested):
There is an issue when I try to pull to refresh with a search field activated but empty:


https://github.com/user-attachments/assets/b65a75a4-b66a-4d52-9771-19020a647e3f

hichamboushaba committed:
Ensure we use some delay when the query gets emptied

To solve an issue where the loading state changes are conflated, resulting in a stuck refresh indicator in the screen.

hichamboushaba committed:
Refactor refreshing logic to avoid nesting flow collection

Also fixes an issue where pull-to-refresh action needed to wait for current fetch to finish before triggering.

hichamboushaba committed:
Fix issue with `take` operator

AdamGrzybkowski reviewed (approved):
:shipit:
